[Kineto] Enable OOM observer

mzzchy opened the pull request:
Summary:
# Context:
When memory leak happens, it usually trigger the OOM in the later iterations. The snapshot of full iteration will be huge and hard to interpret. 
On CUDA side, they provide OOM observer which generates snapshot when OOM happens with latest 1,500,000 entries for debugging.

In this diff, we want to implement the feature on MTIA side

Test Plan:
Run this test with last diff in the stack. 
```
buck run @//mode/opt  kineto/libkineto/fb/mtia/integration_tests:mtia_memory_auto_trace_test
```

As shown, the memory_snapshot is generated when oom happens 
Log: P1794792326
Snapshot: https://fburl.com/pytorch_memory_visualizer/lx73y6s3 {F1977402355}

Differential Revision: D71993315

mzzchy commented:
@pytorchbot label "topic: not user facing"

Skylion007 commented on torch/csrc/mtia/Module.cpp:
```diff
@@ -100,6 +100,11 @@ void initModule(PyObject* module) {
     return py::reinterpret_steal<py::object>(raw_pyobject);
   });
 
+  m.def("_mtia_attachOutOfMemoryObserver", [](py::function observer) {
```
nit: const ref insted of pass by value here??

mzzchy committed:
[Kineto] Enable OOM observer (#152160)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/152160

# Context:
When memory leak happens, it usually trigger the OOM in the later iterations. The snapshot of full iteration will be huge and hard to interpret.
On CUDA side, they provide OOM observer which generates snapshot when OOM happens with latest 1,500,000 entries for debugging.

In this diff, we want to implement the feature on MTIA side

Test Plan:
Run this test with last diff in the stack.
```
buck run @//mode/opt  kineto/libkineto/fb/mtia/integration_tests:mtia_memory_auto_trace_test
```

As shown, the memory_snapshot is generated when oom happens
Log: P1794792326
Snapshot: https://fburl.com/pytorch_memory_visualizer/lx73y6s3 {F1977402355}

Differential Revision: D71993315

sraikund16 reviewed (approved):
LGTM

Skylion007 commented:
@pytorchbot merge

Skylion007 commented:
@pytorchbot merge -f "remaining check flaky"
